### ISSUE TITLE
Container stop timeout

### ISSUE THREAD

**Issue by topjor** [intelwolf/p1monitor#3]
Currently when stopping the container docker will wait 10 seconds by default for a container to stop all processes, and if this timeout is exceeded, will kill it forcefully.
This 10 second timeout is not enough time to stop the container gracefully, as the stop script sleeps 3 or 5 seconds for each process. (current shutdown has 44 seconds of sleep (excluding time needed to run the actual stopping))

In the docker-compose `stop_grace_period` should be set to reflect the longer timeout needed to gracefully exit, or `--stop-timeout` argument for the docker cli.
it might also be useful to mention in your blog post linked in the readme.

I have also noticed that when extending the timeout way to much then it needs to stop, it still takes the full timeout before the stop command returns, suggesting that even after stopping p1monitor gracefully the main process does not stop. so the trap might not break the sleep loop. though I have not confirmed this properly.

**Comment by intelwolf** [comment]
When stopping the container the original script (from the author of p1monitor on Raspberry PI) is invoked. That script is killing the processes one by one and waiting in-between the tasks. In my opinion it does not really matter whether the container is just killed (and all process are stopped automatically) or all process are stopped one by one and the the container is stopped. 
I can check of creating a different stop script that kill the processes in a more speedy way.

**Comment by topjor** [comment]
the problem is two fold:
1. When the stop signal is received the main process invokes p1's original stop command, this is good. but when that command finishes the sleep command does not get broken and the container keeps running doing nothing until docker just kills it forcefully.
2. When docker forcefully kills the container after 10 seconds (the default), the original p1 stop logic is still running and in the middle of committing from ramdisk to storage. resulting in (partial) information loss.

The first problem is not a big of a deal, docker killing the container forcefully, though it does mean that when updating the container, it has to wait longer before it can recreate and roll out the new version, also when gracefully shutting down or restarting the host, it takes longer for no reason.  
But theoretically has no adverse effects. (given that the kill timeout would be bigger then the duration of p1's normal shutdown)

The second problem is a much bigger problem, the loss of information due to killing in the middle of gracefully stopping all processes. Which can result in full corruption even of the retained data.
On a raspberry pi running the container that has been up for just under an hour, it takes longer then 10 seconds to shutdown and write out the database from ramdisk to storage. resulting in partial data retention between last time it synced and shutdown (which is every 15 minutes or every 30 minutes, not sure which one it was)
That was why I proposed to add these parameters to the commands / composer file, and mention it in the readme, so it retains data correctly and shuts down gracefully as long as it is not misbehaving.

As for making it do the shutdown faster, keep in mind that some operations take time, we don't want to kill it anyway in the middle of a process unless its hanging or not working correctly anymore.
Though restructuring the shutdown to be more efficient and faster might me something to take upstream as well.

**Comment by intelwolf** [comment]
If you take a look at the original stop script you'll see that there is actually no action of copying databases from ram to data (talking about how the container is shut down now). The stop action is just killing processes without looking at any current action in progress. Basically the data between the last sync to data and the time of shutting down the container is lost (15 minutes interval). 
I was already looking into invoking a copy action before shutting down. But there is a risk to that because if it is not finished on time it will be corrupted indeed. That is probably only an issue if you are running on a Raspberry Pi though.  So at this time the problem you are describing can possibly only happen if you shut down the container at the moment a database sync is in progress and processes are killed

**Comment by topjor** [comment]
What your entry script does is it sets a trap, capturing stop signals, then you invoke P1's stop script.
The stop script of P1 will then send stop signals to each process in sequence, waiting 3 or 5 seconds, then killing it forcefully.
Each process of P1 also has a `trap` defined to capture those stop signals, and stops its process gracefully. finishing its last task and then stopping, or doing something special like database syncing.
One of the processes (the database one) is for syncing ramdisk to storage. normally doing this once every 15 minutes. When that process receives the stop signal, it will invoke one last sync, so that data does not get lost.

As it currently stands, because of these timeouts when stopping and the force kill from docker after 10 seconds (by default), it will forcefully stop before this copy operation. The database stop job gets called at 14 seconds into the stopping process.

database capturing of stop signal: 
https://github.com/intelwolf/p1monitor/blob/8007465c03365b043623bc706899f2b8701f8fbe/p1mon/scripts/P1Db.py#L1096-L1097

exit function that triggers one last backup:
https://github.com/intelwolf/p1monitor/blob/8007465c03365b043623bc706899f2b8701f8fbe/p1mon/scripts/P1Db.py#L733-L737

**Comment by intelwolf** [comment]
I see what you mean indeed, I looked over the SIGINT in the script. I can of course crate a different stop script but as said if I start copying data earlier in the stop cycle the risk of corruption will increase. So I think adding a stop_grace_signal in that case is the only option. As you mention the sleep actions on its own already take approx 40 seconds so adding the correct period would be at least 1 minute I would say. Haven't checked if the loop breaks but will check on that.
Thanks for the input and suggestions welcome :-)

**Comment by topjor** [comment]
Configuring docker to do the kill later is the only real solution, as otherwise you would have to maintain a fork of the stop functionality, and I take for granted that those timeouts are somewhat correctly set to what may be needed on lower end hardware. You might get away with a little less for some of the processes but some just take time, like the database writing.

I have set my own container to use a stop delay of 180 seconds at first, and I have noticed that it now does correctly write out the database (all processes are stopped in around 35 seconds, but the container keeps running until 180 seconds and the main process gets killed as the trap does not have an `exit 0` at the end, so keeps running.

Actually we don't need more then 45 seconds. As it sends a shutdown to each process, then waits 3 seconds, then kills anyway if still running.
The total was 44 seconds If I calculated correctly. you do still need to add the time it takes to execute the sending of the signal, but that would be negligible so adding 1 or 2 seconds should be fine, unless the system is completely overloaded, but then it would highly likely have to kill anyway.

As for data corruption, that would only happen when it gets killed in the middle of writing out. so by default around 10 seconds, but when this delay gets properly set, should not matter anymore as it gets the time it needs. Corruption only happens when the system is hanging in the middle of writing, a user can recover using the backups that it keeps, losing around 15 to 30 minutes of data then.

**Comment by intelwolf** [comment]
I noticed that the configured `stop_grace_period` will be fully utilised so already also modified the entry script and the container is now terminated when the stop script is completely done or after the timeout set by the `stop_grace_period`.  I also noticed it is about 45 seconds, when setting it to 60 seconds the script has time enough to complete. I'll also modify the documentation accordingly. A new image is already published (1.7.0b)
Thanks for the input
